source/agora/test/RestoreSCPState.d

```diff
@@ -0,0 +1,159 @@
+/*******************************************************************************
+
+    Tests restoring SCP Envelope state on restart
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.RestoreSCPState;
+
+version (unittest):
+
+import agora.common.Config;
+import agora.common.Serializer;
+import agora.common.Task;
+import agora.common.Hash;
+import agora.common.crypto.Key;
+import agora.consensus.protocol.Nominator;
+import agora.consensus.data.Block;
+import agora.consensus.data.ConsensusParams;
+import agora.consensus.data.Transaction;
+import agora.consensus.SCPEnvelopeStore;
+import agora.node.Ledger;
+import agora.node.Validator;
+import agora.network.Clock;
+import agora.network.NetworkManager;
+import agora.test.Base;
+
+import geod24.Registry;
+
+import scpd.Cpp;
+import scpd.types.Stellar_SCP;
+
+import core.stdc.time;
+import core.thread;
+
+/// A test to store SCP state and recover SCP state
+unittest
+{
+    // static variable for comparing values
+    static SCPEnvelope[] envelopes;
+
+    static class ReNominator : TestNominator
+    {
+        /// Ctor
+        public this (immutable(ConsensusParams) params, Clock clock,
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman, string data_dir, ulong txs_to_nominate)
+        {
+            super(params, clock, network, key_pair, ledger, taskman, data_dir,
+                txs_to_nominate);
+        }
+
+        ///
+        override void restoreSCPState ()
+        {
+            //assert(this.scp_envelope_store.length > 0);
```

Comment:
The reason it fails is because the node instantiates a new in-memory database when it restarts. It's empty.

Reply:
Yes, that's right.
I need to keep an instance.

Reply:
And `libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: Cannot set state after starting ballot protocol`
Oh!...

Reply:
I don't know if the ballot protocol started.
I think need to make an exception.

Reply:
I don't see this error anywhere in the CI.

Reply:
And, you can only restore SCP state before nominations, before the event loop starts.

Reply:
I checked the message after testing it locally with FileDB.
I will keep the DB instance above and try to reproduce it.